File: Ruby/lib/civitas-ruby/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require 'io/console'
require_relative 'salidas_carcel'
require_relative 'respuestas'
require_relative 'gestiones_inmobiliarias'

module Civitas

  class Vista_textual

    def initialize
      @i_gestion
      @i_propiedad
      @juegoModel
      @lista_respuestas = [Respuestas::NO, Respuestas::SI]
      @lista_salidas_carcel = [SalidasCarcel::PAGANDO, SalidasCarcel::TIRANDO]
      @lista_gestiones_inmobiliarias = [GestionesInmobiliarias::VENDER, GestionesInmobiliarias:: HIPOTECAR, GestionesInmobiliarias::CANCELAR_HIPOTECA, GestionesInmobiliarias::CONSTRUIR_CASA, GestionesInmobiliarias::CONSTRUIR_HOTEL, GestionesInmobiliarias::TERMINAR]
    end
    
    attr_reader :lista_respuestas, :lista_salidas_carcel, :lista_gestiones_inmobiliarias, i_gestion, i_propiedad

    def mostrar_estado(estado)
      puts estado
    end

    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end

    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l.to_s
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end
    
    def comprar
      lista = @lista_respuestas
      opcion = menu("¿Desea comprar la propiedad actual?", lista)
      opcion
    end

    def gestionar
      lista = @lista_gestiones_inmobiliarias
      @i_gestion = menu("¿Qué desea hacer con la propiedad?", lista)
      @i_propiedad = lee_entero(@juegoModel.jugador_actual.propiedades.size, "\n  Indique el índice de la propiedad que desea gestionar: ",
      "  Valor erróneo") if @i_gestion != 5
    end

    def mostrarSiguienteOperacion(operacion)
      puts operacion
    end

    def mostrarEventos
      while Diario.instance.eventos_pendientes
        puts Diario.instance.leer_evento
      end
    end

    def salir_carcel
      lista = @lista_salidas_carcel
      opcion = menu("Forma de salir de la cárcel: ", lista)
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         #self.actualizarVista
    end

    def actualizarVista
      puts "Jugador actual: 
      #{@juegoModel.jugador_actual.to_s}
Casilla actual: 
      #{@juegoModel.casilla_actual.to_s}
      "
    end
    
  end
end

File: juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Respuestas;

public class VistaTextual {
  
  private CivitasJuego juegoModel; 
  private int iGestion=-1;
  private int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {
    int opcion = menu ("¿Desea comprar la propiedad actual?", 
      new ArrayList<>(Arrays.asList("No", "Sí")));
    return (Respuestas.values()[opcion]);
  }

  void gestionar () {
    iGestion = menu ("¿Qué gestión quiere realizar en sus propiedades?", 
      new ArrayList<> (Arrays.asList("Vender", "Hipotecar", "Cancelar hipoteca", "Construir una casa", "Construir un hotel", "Terminar")));
    if (iGestion != 5) {
      iPropiedad = leeEntero(juegoModel.getJugadorActual().getNumPropiedades(), "\n"+"  Indique el índice de la propiedad a gestionar: ", "  Valor erróneo"); 
    }
  }
  
  public int getGestion(){
    return iGestion;
  }
  
  public int getPropiedad(){
    return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
    System.out.println(operacion);
  }

  void mostrarEventos() {
    while (Diario.getInstance().eventosPendientes()) {
      System.out.println(Diario.getInstance().leerEvento());
    }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
    juegoModel=civitas;
    //this.actualizarVista();
  }
  
  void actualizarVista(){
    System.out.println(separador);
    System.out.println("Jugador actual: " + juegoModel.getJugadorActual().toString() + "\n" +
    "\nCasilla actual: " + juegoModel.getCasillaActual().toString());
    System.out.println(separador);
  } 
}
